### PR TITLE
feat(x/ecocredit): add update basket curator

### PR DESCRIPTION
**Pull request by aleem1314**
## Description

Closes: #1457 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/ecocredit/basket/v1/tx.proto [hunk 1]
```diff
@@ -181,3 +186,26 @@ message MsgUpdateBasketFee {
...
+//
+// Since Revision 2
+message MsgUpdateCurator {
+  option (cosmos.msg.v1.signer) = "curator";
+
+  // denom is the unique identifier of the basket.
+  string denom = 1;
+
+  // curator is the address of the basket curator.
+  string curator = 2;
```
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1/tx.proto:200]: nit: signer is usually the 1st field

File: x/ecocredit/basket/keeper/msg_update_curator_test.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	err := json.Unmarshal([]byte(a.Content), &basket)
+	require.NoError(s.t, err)
+
+	err = s.k.stateStore.BasketTable().Save(s.ctx, &api.Basket{
+		BasketDenom:      basket.BasketDenom,
+		Name:             basket.Name,
+		CreditTypeAbbrev: basket.CreditTypeAbbrev,
+		Curator:          s.alice,
+	})
+	require.NoError(s.t, err)
```
Comment (technicallyty) [inline x/ecocredit/basket/keeper/msg_update_curator_test.go:46]: recently discovered this issue https://github.com/cosmos/cosmos-sdk/issues/12332

i dont think its really necessary to change the tests here in this PR, unless you want to, but just adding a note here that we probably want to switch to gotest.tools/v3/assert for future tests. tracking: https://github.com/regen-network/regen-ledger/issues/1487

File: x/ecocredit/basket/types/v1/features/msg_update_curator.feature [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    Given the message
+	"""
+	{
+		"curator": "regen1ua97smk5yv26wpqmftgdg0sx0q0d38vky7998g",
+		"new_curator": "regen1elq7ys34gpkj3jyvqee0h6yk4h9wsfxmgqelsw",
+		"denom": "eco.uC.NCT"
+	}
+	"""
+	When the message is validated
+    Then expect no error
```
Comment (technicallyty) [inline x/ecocredit/basket/types/v1/features/msg_update_curator.feature:13]: formatting

**Review by technicallyty — APPROVED** [review]
LGTM! just a few comments

File: x/ecocredit/basket/keeper/features/msg_update_curator.feature [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    Scenario: the basket curator is updated
+      When alice attempts to update basket curator with denom "eco.uC.NCT"
+      Then expect basket with denom "eco.uC.NCT" and curator bob
+      And expect event with properties
+      """
+      {
+        "denom": "eco.uC.NCT"
+      }
+      """
```
Comment (ryanchristo) [inline x/ecocredit/basket/keeper/features/msg_update_curator.feature:86]:
```suggestion
      """

```

File: x/ecocredit/basket/types/v1/msg_update_curator.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	if _, err := sdk.AccAddressFromBech32(m.NewCurator); err != nil {
+		return sdkerrors.ErrInvalidAddress.Wrapf("new curator: %s", err)
+	}
+
+	if m.NewCurator == m.Curator {
+		return sdkerrors.ErrInvalidAddress.Wrap("curator and new curator cannot be the same")
+	}
+
+	if err := basket.ValidateBasketDenom(m.Denom); err != nil {
+		return err
```
Comment (ryanchristo) [inline x/ecocredit/basket/types/v1/msg_update_curator.go:39]: Might be nice to wrap the parse error in an invalid request error for consistency. Not sure if this is how we want to do things in the long run but we should be consistent for now. See https://github.com/regen-network/regen-ledger/issues/1147#issuecomment-1199793577 for reference.

File: x/ecocredit/basket/keeper/msg_update_curator_test.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+func (s *updateBasketCurator) Before(t gocuke.TestingT) {
+	s.baseSuite = setupBase(t)
+	s.alice = s.addrs[0]
+	s.bob = s.addrs[1]
+}
+
+func (s *updateBasketCurator) ABasketWithPropertiesAndCuratorAlice(a gocuke.DocString) {
+	var basket types.Basket
+	err := json.Unmarshal([]byte(a.Content), &basket)
```
Comment (ryanchristo) [inline x/ecocredit/basket/keeper/msg_update_curator_test.go:37]:
```suggestion
	err := jsonpb.UnmarshalString(a.Content, &basket)
```

Comment (ryanchristo) [reply x/ecocredit/basket/keeper/msg_update_curator_test.go:37]: We should be using `jsonpb` when we can. Will need to update the import as well.

File: x/ecocredit/basket/keeper/msg_update_curator_test.go (hunk 2 above)
Comment (ryanchristo) [reply x/ecocredit/basket/keeper/msg_update_curator_test.go:46]: Oof. I do like the methods available in require but makes sense. There is a command line tool to automate the migration so we can handle all at once in a separate pull request. Also see https://github.com/regen-network/regen-ledger/issues/1487#issuecomment-1247343297

**Review by ryanchristo — APPROVED** [review]
[tACK](https://hackmd.io/qF_IoJ5eQJ2K6g2Zmb3XCw)

**Pull request merged**